libcudacxx/include/cuda/std/__cmath/fpclassify.h

```diff
@@ -0,0 +1,189 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___CMATH_FPCLASSIFY_H
+#define _LIBCUDACXX___CMATH_FPCLASSIFY_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__bit/bit_cast.h>
+#include <cuda/std/__cmath/common.h>
+#include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__type_traits/is_extended_floating_point.h>
+#include <cuda/std/__type_traits/is_integral.h>
+
+#if _CCCL_COMPILER(NVRTC)
+#  ifndef FP_NAN
+#    define FP_NAN 0
+#  endif // ! FP_NAN
+#  ifndef FP_INFINITE
+#    define FP_INFINITE 1
+#  endif // ! FP_INFINITE
+#  ifndef FP_ZERO
+#    define FP_ZERO 2
+#  endif // ! FP_ZERO
+#  ifndef FP_SUBNORMAL
+#    define FP_SUBNORMAL 3
+#  endif // ! FP_SUBNORMAL
+#  ifndef FP_NORMAL
+#    define FP_NORMAL 4
+#  endif // ! FP_NORMAL
+#else // ^^^ _CCCL_COMPILER(NVRTC) ^^^ ^/  vvv !_CCCL_COMPILER(NVRTC) vvv
+#  include <math.h>
+#endif // !_CCCL_COMPILER(NVRTC)
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+struct _CCCL_FLOAT_BITS
+{
+#if defined(_LIBCUDACXX_LITTLE_ENDIAN)
+  unsigned int man  : 23;
+  unsigned int exp  : 8;
+  unsigned int sign : 1;
+#else // ^^^ _LIBCUDACXX_LITTLE_ENDIAN ^^^ / vvv _LIBCUDACXX_BIG_ENDIAN vvv
+  unsigned int sign : 1;
+  unsigned int exp  : 8;
+  unsigned int man  : 23;
+#endif // _LIBCUDACXX_BIG_ENDIAN
+};
```

Comment:
AFAIK, the layout of bitfields is [implementation defined](https://eel.is/c++draft/class.bit#1):

> Allocation of bit-fields within a class object is implementation-defined. Alignment of bit-fields is implementation-defined.

> Note 1: Bit-fields straddle allocation units on some machines and not on others. Bit-fields are assigned right-to-left on some machines, left-to-right on others — end note

I don't think we can portably reinterpret an FP type as a bitfield :S